lib/gvm/SubstrateStaticJNA.java

```diff
@@ -0,0 +1,196 @@
+     * @param filename Expected filename for the library.
+     * @return The extracted library file.
+     */
+    private static File unpackLibrary(String resource, String filename) {
+        // Unpack the static library from resources so Native Image can use it
+        File extractedLib;
+        try {
+            extractedLib = Native.extractFromResourcePath(resource, Native.class.getClassLoader());
+
+            // The library is extracted into a file with a `.tmp` name, which will not be picked up by the linker
+            // We need to rename it first using the platform-specific convention or the build will fail
+            File platformLib = new File(extractedLib.getParentFile(), filename);
+            if (!extractedLib.renameTo(platformLib)) throw new IllegalStateException("Renaming extract file failed");
+            extractedLib = platformLib;
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to extract native dispatch library from resources", e);
+        }
+        return extractedLib;
+    }
+
+    @Override
+    public String getDescription() {
+        return "Enables optimized static access to JNA at runtime";
+    }
+
+    @Override
+    public boolean isInConfiguration(IsInConfigurationAccess access) {
+        return access.findClassByName(JavaNativeAccess.NATIVE_LAYOUT) != null;
+    }
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        var nativeLibraries = NativeLibrarySupport.singleton();
+        var platformLibraries = PlatformNativeLibrarySupport.singleton();
+
+        // Register as a built-in library with Native Image and set the name prefix used by native symbols
+        nativeLibraries.preregisterUninitializedBuiltinLibrary(JNA_LINK_NAME);
+        platformLibraries.addBuiltinPkgNativePrefix(JNA_NATIVE_LAYOUT);
+
+        // Extract the main JNA library from the platform-specific resource path; next, extract the FFI
+        // library it depends on
+        unpackLibrary(getFFILibraryResource(), getFFILibraryFileName());
+        var extractedLib = unpackLibrary(getStaticLibraryResource(), getStaticLibraryFileName());
+
+        // WARNING: the static JNI linking feature is unstable and may be removed in the future;
+        // this code uses the access implementation directly in order to register the static library. We
+        // inform the Native Image compiler that JNA depends on `ffi`, so that it forces it to load first
+        // when JNA is initialized at image runtime.
+        var nativeLibsImpl = ((BeforeAnalysisAccessImpl) access).getNativeLibraries();
+        nativeLibsImpl.addStaticNonJniLibrary(FFI_LINK_NAME);
+        nativeLibsImpl.addStaticJniLibrary(JNA_LINK_NAME, FFI_LINK_NAME);
+
+        // Enhance the Native Image lib paths so the injected static libraries are available to the linker
+        nativeLibsImpl.getLibraryPaths().add(extractedLib.getParentFile().getAbsolutePath());
```

Comment:
Linking would fail before it was added, the linker couldn't find the symbols for `jnidispatch`, though a few things have changed now, definitely should re-test without it.

Reply:
i mean whatever